include/aws/s3/private/s3_checksums.h

```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_S3_Checksums_H
+#define AWS_S3_Checksums_H
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "aws/s3/s3_client.h"
+#include <aws/cal/hash.h>
+#include <aws/io/stream.h>
+
+#define AWS_CRC32_LEN 4
+#define AWS_CRC32C_LEN 4
+
+struct aws_checksum;
+
+struct aws_checksum_vtable {
+    void (*destroy)(struct aws_checksum *checksum);
+    int (*update)(struct aws_checksum *checksum, const struct aws_byte_cursor *buf);
+    int (*finalize)(struct aws_checksum *checksum, struct aws_byte_buf *out, size_t truncate_to);
+};
+
+struct aws_checksum {
+    struct aws_allocator *allocator;
+    struct aws_checksum_vtable *vtable;
+    void *impl;
+    size_t digest_size;
+    bool good;
+};
+
+AWS_S3_API
+struct aws_input_stream *aws_checksum_stream_new(
+    struct aws_allocator *allocator,
+    struct aws_input_stream *existing_stream,
+    enum aws_s3_checksum_algorithm algorithm,
+    struct aws_byte_buf *checksum_result);
+
+AWS_S3_API
+struct aws_input_stream *aws_chunk_stream_new(
+    struct aws_allocator *allocator,
+    struct aws_input_stream *existing_stream,
+    enum aws_s3_checksum_algorithm algorithm);
+
+AWS_S3_API
+size_t digest_size_from_algorithm(enum aws_s3_checksum_algorithm algorithm);
+
+/**
+ * Allocates and initializes a sha256 checksum instance.
+ */
+AWS_S3_API struct aws_checksum *aws_sha256_checksum_new(struct aws_allocator *allocator);
+/**
+ * Allocates and initializes a sha1 checksum instance.
+ */
+AWS_S3_API
+struct aws_checksum *aws_sha1_checksum_new(struct aws_allocator *allocator);
+/**
+ * Allocates and initializes an md5 checksum instance.
+ */
+AWS_S3_API
+struct aws_checksum *aws_md5_checksum_new(struct aws_allocator *allocator);
+/**
```

Comment:
Minor preference but I really like it when we put a newline between declaration-comment blocks.